src/finch/compiled.py

```diff
@@ -2,10 +2,66 @@
 from functools import wraps
 
 from .julia import jl
-from .tensor import Tensor
 from .typing import JuliaObj
 
 
+def _recurse(x, /, *, f):
+    if isinstance(x, tuple | list):
+        return type(x)(_recurse(xi, f=f) for xi in x)
+    if isinstance(x, dict):
+        return {k: _recurse(v, f=f) for k, v in x}
+    return f(x)
+
+
+def _recurse_iter(x, /):
+    if isinstance(x, tuple | list):
+        yield from (_recurse_iter(xi) for xi in x)
+    if isinstance(x, dict):
+        yield from (_recurse_iter(xi) for xi in x.values())
+    yield x
+
+
+def _to_lazy_tensor(x, /):
+    from .tensor import Tensor
+
+    if isinstance(x, Tensor) and not jl.isa(x._obj, jl.Finch.LazyTensor):
+        return Tensor(jl.Finch.LazyTensor(x._obj))
+
+    return x
+
+
+def compiled(opt=None, *, force_materialization=False, tag: int | None = None):
```

Comment:
As it's a new feature I would like to see a few tests for it:
- Test with a nested `@compiled` decorators.
- Test with `force_materialization=True` which checks that for all lazy inputs the output is computed.

Reply:
Done in https://github.com/finch-tensor/finch-tensor-python/pull/83/commits/4d45756fa3ed78c2c8785a15c81ffeda94d5cac8.